wfe2/wfe.go

```diff
@@ -2130,13 +2202,38 @@ func (wfe *WebFrontEndImpl) NewOrder(
 
 	logEvent.DNSNames = names
 
+	// TODO(#5545): Spending and Refunding can be async until these rate limits
+	// are authoritative. This saves us from adding latency to each request.
+	doneCheckingLimits := make(chan struct{})
+	var ratelimitTxns []ratelimits.Transaction
+	var newOrderSuccessful bool
+	var errIsRateLimit bool
+
+	go func() {
+		// Close the channel on goroutine completion.
+		defer close(doneCheckingLimits)
+		ratelimitTxns = wfe.checkNewOrderLimits(ctx, acct.ID, names)
+	}()
+
+	defer func() {
+		// Wait for the rate limit check to complete before attempting to refund
+		// the limits. If the check failed, we don't want to refund anything.
+		<-doneCheckingLimits
+		if !newOrderSuccessful && !errIsRateLimit && ratelimitTxns != nil {
```

Comment:
Is it actually a problem to call BatchRefund(nil)? I believe that should be a fast no-op, and would allow you to remove the extra condition here.